package.json

```diff
@@ -1,30 +1,39 @@
 {
-  "name": "pname",
+  "name": "worker-typescript-template",
   "version": "1.0.0",
   "description": "Cloudflare worker TypeScript template",
-  "main": "index.js",
+  "main": "src/index.ts",
   "scripts": {
     "build": "webpack",
-    "dev": "NODE_ENV=development npm run build",
-    "format": "prettier --write '**/*.{ts,js,css,json,md}'",
-    "test:clean": "rimraf ./transpiled/src ./transpiled/test",
-    "test": "npm run test:clean && npm run transpile && mocha --require source-map-support/register --recursive transpiled/test",
-    "transpile": "tsc --project ./test"
+    "format": "prettier --write  '*.{json,js}' 'src/**/*.{js,ts}' 'test/**/*.{js,ts}'",
+    "lint": "eslint --max-warnings=0 src && prettier --check '*.{json,js}' 'src/**/*.{js,ts}' 'test/**/*.{js,ts}'",
+    "test": "jest --config jestconfig.json --verbose"
   },
   "author": "author",
   "license": "MIT OR Apache-2.0",
-  "devDependencies": {
-    "@cloudflare/workers-types": "^2.0.0",
-    "@types/chai": "^4.2.11",
-    "@types/mocha": "^7.0.2",
-    "chai": "^4.2.0",
-    "mocha": "^8.0.1",
-    "prettier": "^2.0.5",
-    "rimraf": "^3.0.2",
+  "eslintConfig": {
+    "root": true,
+    "extends": [
+      "typescript",
+      "prettier"
+    ]
+  },
+  "dependencies": {
+    "@cloudflare/workers-types": "^2.2.2",
+    "@types/jest": "^26.0.23",
+    "@types/service-worker-mock": "^2.0.1",
+    "@typescript-eslint/eslint-plugin": "^4.16.1",
+    "@typescript-eslint/parser": "^4.16.1",
+    "eslint": "^7.21.0",
+    "eslint-config-prettier": "^8.1.0",
+    "eslint-config-typescript": "^3.0.0",
+    "jest": "^27.0.1",
+    "prettier": "^2.3.0",
     "service-worker-mock": "^2.0.5",
-    "ts-loader": "^7.0.5",
-    "typescript": "^3.9.5",
-    "webpack": "^4.43.0",
-    "webpack-cli": "^3.3.12"
+    "ts-jest": "^27.0.1",
+    "ts-loader": "^8",
+    "typescript": "^4.2.4",
```

Comment:
This needs to be changed to `"typescript": "^3"`.

Basically dependencies work something like:
* `wrangler` doesn't work with `webpack` 5, only 4
* `webpack` doesn't work with `ts-loader` 9, only 8
* `ts-loader` doesn't work with `typescript` 4, only 3 (this is not documented)

So currently you can't use ts4, you need to have:
* webpack 5
* ts-loader 8
* typescript 3

Reply:
What exactly were you trying to do, and what errors were you getting?

Reply:
I was trying to use the `"jsx": "react-jsx"` typescript feature which is only available in typescript 4.

The errors are different for each case above, and TBH I can't remember the exact error for each one. Best you setup a project and try them yourself.

Reply:
I did setup a project before asking you what errors you were getting. Adding dependencies, importing from them, and using typescript 4 features such as variadic tuple types appears to work fine with wrangler build and wrangler publish, and the resulting worker behaves as I'd expect.

Can you share a minimal repro case of the errors you were seeing?

Reply:
Okay, below are the errors for each case, I've also created a repo [here](https://github.com/samuelcolvin/wrangler-typescript-errors). The 3 commits represent each of the three cases.

## `ts-loader` doesn't work with `typescript` 4, only 3 (this is not documented)

deps:
```
    "ts-loader": "^8",
    "typescript": "^4.2.4",
    "webpack": "^4",
```
error:
```
demo-ts 0  11ms ➤  yarn build
yarn run v1.22.10
$ webpack
Hash: aa5f0adf9f51c59c4264
Version: webpack 4.46.0
Time: 621ms
Built at: 02/06/2021 16:13:58
        Asset      Size  Chunks                   Chunk Names
    worker.js  5.21 KiB    main  [emitted]        main
worker.js.map  4.05 KiB    main  [emitted] [dev]  main
Entrypoint main = worker.js worker.js.map
[./src/handler.ts] 427 bytes {main} [built]
[./src/index.ts] 221 bytes {main} [built]
[./src/utils.ts] 350 bytes {main} [built] [failed] [1 error]

ERROR in ./src/utils.ts 5:10
Module parse failed: Unexpected token (5:10)
File was processed with these loaders:
 * ./node_modules/ts-loader/index.js
You may need an additional loader to handle the result of these loaders.
| exports.json_response = exports.HttpError = void 0;
| class HttpError extends Error {
>     status;
|     body;
|     headers;
 @ ./src/handler.ts 4:16-34
 @ ./src/index.ts
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```


## `webpack` doesn't work with `ts-loader` 9, only 8

deps:
```
    "ts-loader": "^9.2.2",
    "typescript": "^4.2.4",
    "webpack": "^4",
```
error:
```
demo-ts 0 2.77s ➤  yarn build
yarn run v1.22.10
$ webpack
Hash: 4835201e14a661bdfbc6
Version: webpack 4.46.0
Time: 70ms
Built at: 02/06/2021 16:15:18
        Asset      Size  Chunks                   Chunk Names
    worker.js  4.07 KiB    main  [emitted]        main
worker.js.map  3.47 KiB    main  [emitted] [dev]  main
Entrypoint main = worker.js worker.js.map
[./src/index.ts] 325 bytes {main} [built] [failed] [1 error]

ERROR in ./src/index.ts
Module build failed (from ./node_modules/ts-loader/index.js):
TypeError: loaderContext.getOptions is not a function
    at getLoaderOptions (/home/samuel/code/demo-ts/node_modules/ts-loader/dist/index.js:91:41)
    at Object.loader (/home/samuel/code/demo-ts/node_modules/ts-loader/dist/index.js:14:21)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

## `wrangler` doesn't work with `webpack` 5, only 4

deps:
```
    "ts-loader": "^9.2.2",
    "typescript": "^4.2.4",
    "webpack": "^5.38.1",
```
error:
```
demo-ts 130 17.80s ➤  wrangler preview
👀  ./src/index.ts
Module build failed (from ./node_modules/ts-loader/index.js):
TypeError: loaderContext.getOptions is not a function
    at getLoaderOptions (/home/samuel/code/demo-ts/node_modules/ts-loader/dist/index.js:91:41)
    at Object.loader (/home/samuel/code/demo-ts/node_modules/ts-loader/dist/index.js:14:21)
Error: webpack returned an error. Try configuring `entry` in your webpack config relative to the current working directory, or setting `context = __dirname` in your webpack config.
```
(weirdly `yarn build` (e.g. `webpack`) works fine in this case)

Reply:
Thanks for the repro.

Pushed a version of this template using custom builds.

Copying the src directory from master of your repo into the template, wrangler publish works. Please confirm.